oidc-controller/api/services/cleanup.py

```diff
@@ -0,0 +1,137 @@
+"""Background cleanup service for presentation records."""
+
+import asyncio
+from datetime import datetime, timedelta, UTC
+from typing import List
+
+import structlog
+
+from ..core.config import settings
+from ..core.acapy.client import AcapyClient
+
+logger: structlog.typing.FilteringBoundLogger = structlog.getLogger(__name__)
+
+
+class PresentationCleanupService:
+    """Service for cleaning up old presentation records."""
+
+    def __init__(self):
+        self.client = AcapyClient()
+        self.retention_hours = settings.CONTROLLER_PRESENTATION_RECORD_RETENTION_HOURS
+        self.schedule_minutes = (
+            settings.CONTROLLER_PRESENTATION_CLEANUP_SCHEDULE_MINUTES
+        )
+
+    async def cleanup_old_presentation_records(self) -> dict:
+        """Clean up presentation records older than retention period."""
+        logger.info("Starting background cleanup of old presentation records")
+
+        cutoff_time = datetime.now(UTC) - timedelta(hours=self.retention_hours)
+        cleanup_stats = {
+            "total_records": 0,
+            "cleaned_records": 0,
+            "failed_cleanups": 0,
+            "errors": [],
+        }
+
+        try:
+            # Get all presentation records
+            records = self.client.get_all_presentation_records()
+            cleanup_stats["total_records"] = len(records)
+
+            logger.info(
+                f"Found {len(records)} presentation records for cleanup evaluation"
+            )
+
+            for record in records:
+                try:
+                    # Parse the creation time - ACA-Py uses ISO format
+                    created_at_str = record.get("created_at")
+                    # Parse ISO timestamp (handle both with and without timezone)
+                    try:
+                        if created_at_str.endswith("Z"):
+                            created_at_str = created_at_str[:-1] + "+00:00"
+                        record_time = datetime.fromisoformat(created_at_str)
+                        if record_time.tzinfo is None:
+                            record_time = record_time.replace(tzinfo=UTC)
+                    except ValueError as parse_error:
+                        logger.warning(
+                            f"Failed to parse timestamp {created_at_str} for record {record.get('pres_ex_id')}: {parse_error}"
+                        )
+                        continue
+
+                    # Check if record is old enough to clean up
+                    if record_time < cutoff_time:
+                        pres_ex_id = record.get("pres_ex_id")
+                        presentation_deleted, _, errors = (
+                            self.client.delete_presentation_record_and_connection(
```

Comment:
O(n) API calls over to ACA-Py is... going to be ok for heavy load?

Reply:
Haha. Might be a bit of an experiment. The delete endpoint is very simple look up and delete by indexed key so it should be performant.